File: src/main/scala/org/xbery/artbeams/articles/repository/ArticleCategoryMapper.scala
package org.xbery.artbeams.articles.repository

import org.xbery.artbeams.articles.domain.ArticleCategory
import org.xbery.overview.filter.Condition
import org.xbery.overview.mapper._
import org.xbery.overview.repo.Conditions

import java.util

/**
  * Maps Article-Category binding to set of attributes and vice versa.
  * @author Radek Beran
  */
class ArticleCategoryMapper() extends DynamicEntityMapper[ArticleCategory, ArticleCategoryFilter] {

  private val cls = classOf[ArticleCategory]

  override val getTableName: String = "article_category"

  val articleIdAttr = add(Attr.ofString(cls, "article_id").get(e => e.articleId).primary())
  val categoryIdAttr = add(Attr.ofString(cls, "category_id").get(e => e.categoryId).primary())

  override def createEntity(attributeSource: AttributeSource, attributes: java.util.List[Attribute[ArticleCategory, _]], aliasPrefix: String): ArticleCategory = {
    ArticleCategory(
      articleIdAttr.getValueFromSource(attributeSource, aliasPrefix),
      categoryIdAttr.getValueFromSource(attributeSource, aliasPrefix)
    )
  }

  override def composeFilterConditions(filter: ArticleCategoryFilter): util.List[Condition] = {
    val conditions = new util.ArrayList[Condition]
    filter.articleId.map(articleId => conditions.add(Conditions.eq(this.articleIdAttr, articleId)))
    filter.categoryId.map(categoryId => conditions.add(Conditions.eq(this.categoryIdAttr, categoryId)))
    conditions
  }
}

object ArticleCategoryMapper {
  lazy val Instance = new ArticleCategoryMapper()
}

File: src/main/kotlin/org/xbery/artbeams/articles/repository/ArticleCategoryMapper.kt
package org.xbery.artbeams.articles.repository

import org.xbery.artbeams.articles.domain.ArticleCategory
import org.xbery.overview.filter.Condition
import org.xbery.overview.mapper.Attr
import org.xbery.overview.mapper.Attribute
import org.xbery.overview.mapper.AttributeSource
import org.xbery.overview.mapper.DynamicEntityMapper
import org.xbery.overview.repo.Conditions
import java.util.*

/**
 * Maps Article-Category binding to set of attributes and vice versa.
 * @author Radek Beran
 */
open class ArticleCategoryMapper : DynamicEntityMapper<ArticleCategory, ArticleCategoryFilter>() {
    private val cls: Class<ArticleCategory> = ArticleCategory::class.java
    override fun getTableName(): String = "article_category"
    val articleIdAttr: Attribute<ArticleCategory, String> = add(Attr.ofString(cls, "article_id").get { e -> e.articleId }.primary())
    val categoryIdAttr: Attribute<ArticleCategory, String> =
        add(Attr.ofString(cls, "category_id").get { e -> e.categoryId }.primary() )

    override fun createEntity(
        attributeSource: AttributeSource,
        attributes: List<Attribute<ArticleCategory, *>>,
        aliasPrefix: String?
    ): ArticleCategory {
        return ArticleCategory(articleIdAttr.getValueFromSource(
          attributeSource,
          aliasPrefix
        ), categoryIdAttr.getValueFromSource(attributeSource, aliasPrefix ?: ""))
    }

    override fun composeFilterConditions(filter: ArticleCategoryFilter): MutableList<Condition> {
        val conditions = mutableListOf<Condition>()
        filter.articleId?.let { articleId -> conditions.add(Conditions.eq(this.articleIdAttr, articleId)) }
        filter.categoryId?.let { categoryId -> conditions.add(Conditions.eq(this.categoryIdAttr, categoryId)) }
        return conditions
    }

    companion object {
        val Instance: ArticleCategoryMapper = ArticleCategoryMapper()
    }
}
